### ISSUE TITLE
 Package: python3:x64-windows  issue

### ISSUE THREAD

**Issue by jbmartens** [microsoft/vcpkg#6310]
```
 PS C:\vcpkg-master> .\vcpkg install mlpack[tools]:x64-windows
The following packages will be built and installed:
  * boost[core]:x64-windows
  * boost-parameter-python[core]:x64-windows
  * boost-poly-collection[core]:x64-windows
  * boost-process[core]:x64-windows
  * boost-program-options[core]:x64-windows
  * boost-python[core]:x64-windows
  * boost-sort[core]:x64-windows
  * boost-stacktrace[core]:x64-windows
  * boost-statechart[core]:x64-windows
  * boost-wave[core]:x64-windows
  * boost-yap[core]:x64-windows
  * clapack[core]:x64-windows
    mlpack[core,tools]:x64-windows
  * python3[core]:x64-windows
Additional packages (*) will be modified to complete this operation.
Starting package 1/14: python3:x64-windows
Building package python3[core]:x64-windows...
-- Using cached C:/vcpkg-master/downloads/python-cpython-v3.7.3.tar.gz
-- Extracting source C:/vcpkg-master/downloads/python-cpython-v3.7.3.tar.gz
-- Using source at C:/vcpkg-master/buildtrees/python3/src/v3.7.3-1161ad7a91
-- Building C:/vcpkg-master/buildtrees/python3/src/v3.7.3-1161ad7a91-Lib-dynamic-crt-dynamic/PCBuild/pythoncore.vcxproj for Release
CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:56 (message):
    Command failed: msbuild;C:/vcpkg-master/buildtrees/python3/src/v3.7.3-1161ad7a91-Lib-dynamic-crt-dynamic/PCBuild/pythoncore.vcxproj;/p:Configuration=Release;/t:Rebuild;/p:Platform=x64;/p:PlatformToolset=v140;/p:VCPkgLocalAppDataDisabled=true;/p:UseIntelMKL=No;/p:WindowsTargetPlatformVersion=;/m
    Working Directory: C:/vcpkg-master/buildtrees/python3/x64-windows-rel
    Error code: 1
    See logs for more information:
      C:\vcpkg-master\buildtrees\python3\build-x64-windows-rel-out.log

Call Stack (most recent call first):
  scripts/cmake/vcpkg_build_msbuild.cmake:119 (vcpkg_execute_required_process)
  ports/python3/portfile.cmake:37 (vcpkg_build_msbuild)
  scripts/ports.cmake:71 (include)


Error: Building package python3:x64-windows failed with: BUILD_FAILED
Please ensure you're using the latest portfiles with `.\vcpkg update`, then
submit an issue at https://github.com/Microsoft/vcpkg/issues including:
  Package: python3:x64-windows
  Vcpkg version: 2018.11.23-nohash
```

**Comment by zeFresk** [comment]
I had the same problem it seems boost-python/portfile doesn't work with multiple version of Python installed (?) or that the line 
`file(GLOB PYTHON_INCLUDE_PATH "${CURRENT_INSTALLED_DIR}/include/python[0-9.]*")`
is not working as expected :(
A quick fix that worked for me is to define the include path manually : 
`set(PYTHON_INCLUDE_PATH "C:/vcpkg/installed/x64-windows/include/python3.7")`

**Comment by jbmartens** [comment]
Thanks for the tio. I tried adding

set(PYTHON_INCLUDE_PATH "C:/vcpkg/installed/x64-windows/include/python3.7")

to the ports/boost-python/portfile.cmake  file and rerun

.\vcpkg install mlpack[tools]:x64-windows

but I keep getting the same error message.

I indeed have Anaconda + Python installed, so an alternative could be to remove this installation.

**Comment by zeFresk** [comment]
Did you add it before or after the line with `GLOB`?
I have Anaconda, Python 2 & 3 installed so I don't think that's the problem (with my fix).

**Comment by jbmartens** [comment]
I tried both before and after the GLOB statement (not sure which one is the correct order).
I noticed that the directory  C:/vcpkg/installed/x64-windows/include/python3.7 doesn’t seem to exist.

The feedback that I get is the following:

**Comment by hyseanlin** [comment]
I have Anaconda + Python 3 installed.
I tried the fix proposed by @zeFresk, but had the same problem.
Is there any workaround?

**Comment by jbmartens** [comment]
I tried both before and after the GLOB statement (not sure which one is the correct order).
I noticed that the directory  C:/vcpkg/installed/x64-windows/include/python3.7 doesn’t seem to exist.

The feedback that I get is the following:

PS C:\vcpkg-master> .\vcpkg install mlpack[tools]:x64-windows
The following packages will be built and installed:
  * boost[core]:x64-windows
  * boost-parameter-python[core]:x64-windows
  * boost-poly-collection[core]:x64-windows
  * boost-process[core]:x64-windows
  * boost-program-options[core]:x64-windows
  * boost-python[core]:x64-windows
  * boost-sort[core]:x64-windows
  * boost-stacktrace[core]:x64-windows
  * boost-statechart[core]:x64-windows
  * boost-wave[core]:x64-windows
  * boost-yap[core]:x64-windows
  * clapack[core]:x64-windows
    mlpack[core,tools]:x64-windows
  * python3[core]:x64-windows
Additional packages (*) will be modified to complete this operation.
Starting package 1/14: python3:x64-windows
Building package python3[core]:x64-windows...
-- Using cached C:/vcpkg-master/downloads/python-cpython-v3.7.3.tar.gz
-- Extracting source C:/vcpkg-master/downloads/python-cpython-v3.7.3.tar.gz
-- Using source at C:/vcpkg-master/buildtrees/python3/src/v3.7.3-1161ad7a91
-- Building C:/vcpkg-master/buildtrees/python3/src/v3.7.3-1161ad7a91-Lib-dynamic-crt-dynamic/PCBuild/pythoncore.vcxproj for Release
CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:56 (message):
    Command failed: msbuild;C:/vcpkg-master/buildtrees/python3/src/v3.7.3-1161ad7a91-Lib-dynamic-crt-dynamic/PCBuild/pythoncore.vcxproj;/p:Configuration=Release;/t:Rebuild;/p:Platform=x64;/p:PlatformToolset=v140;/p:VCPkgLocalAppDataDisabled=true;/p:UseIntelMKL=No;/p:WindowsTargetPlatformVersion=;/m
    Working Directory: C:/vcpkg-master/buildtrees/python3/x64-windows-rel
    Error code: 1
    See logs for more information:
      C:\vcpkg-master\buildtrees\python3\build-x64-windows-rel-out.log

Call Stack (most recent call first):
  scripts/cmake/vcpkg_build_msbuild.cmake:119 (vcpkg_execute_required_process)
  ports/python3/portfile.cmake:37 (vcpkg_build_msbuild)
  scripts/ports.cmake:71 (include)


Error: Building package python3:x64-windows failed with: BUILD_FAILED
Please ensure you're using the latest portfiles with `.\vcpkg update`, then
submit an issue at https://github.com/Microsoft/vcpkg/issues including:
  Package: python3:x64-windows
  Vcpkg version: 2018.11.23-nohash

Additionally, attach any relevant sections from the log files above.

**Comment by NancyLi1013** [comment]
Hi @jbmartens thanks for posting this issue.  Does this issue still exist on your machine now?
I can install mlpack on my local successfully.  if this issue can be reproduced now, please provide your log.

Thanks.

**Comment by timautry** [comment]
If more than one python version is installed, the boost-python install fails. If I hard code the path to the version of python needed and remove the search in the portfile.cmake then boost-python package will install properly - windows 10 64-bit Version 1909 (OS Build 18363.476 and the latest updates to VS 2017 and VS 2019 - example of portfile.cmake with hardcoded PYTHON_INCLUDE_PATH:

_# Find Python. Can't use find_package here, but we already know where everything is _
**# file(GLOB PYTHON_INCLUDE_PATH "${CURRENT_INSTALLED_DIR}/include/python[0-9.]*")**
**set(PYTHON_INCLUDE_PATH "D:/Git/vs2017/installed/x64-windows/include/python3.7")**
set(PYTHONLIBS_RELEASE "${CURRENT_INSTALLED_DIR}/lib")

**Comment by JackBoosY** [comment]
Related: #6639 #8882.

**Comment by yurybura** [comment]
It should be fixed with #9317

**Comment by NancyLi1013** [comment]
Hi all,
This issue should be fixed now. 
I can install `mlpack:x64-windows` successfully on my local.
`boost-python` can be installed successfully, too. Even though installed both `python2 `and `python3`.

Please try to update vcpkg and install them again.

**Comment by JackGoldfinch** [comment]
Unfortunately, it seems not to be fixed for me...

If I read this thread correctly, it wasn't really about mlpack or boost-python,  the failing build was python3[core]:x64-windows. Or, in my case, _is_ the failing build.

After updating vcpkg to #9591, I still cannot build python3. During installation, I get almost exactly the same output as the OP:

> The following packages will be built and installed:
    python3[core]:x64-windows
Starting package 1/1: python3:x64-windows
Building package python3[core]:x64-windows...
-- Extracting source C:/Dev/vcpkg/downloads/python-cpython-v3.7.3.tar.gz
-- Using source at C:/Dev/vcpkg/buildtrees/python3/src/v3.7.3-1161ad7a91
-- Building C:/Dev/vcpkg/buildtrees/python3/src/v3.7.3-1161ad7a91-Lib-Win/PCBuild/pythoncore.vcxproj for Release        CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:72 (message):
    Command failed: msbuild C:/Dev/vcpkg/buildtrees/python3/src/v3.7.3-1161ad7a91-Lib-Win/PCBuild/pythoncore.vcxproj /p:Configuration=Release /t:Rebuild /p:Platform=x64 /p:PlatformToolset=v142 /p:VCPkgLocalAppDataDisabled=true /p:UseIntelMKL=No /p:WindowsTargetPlatformVersion=10.0.18362.0 /m
    Working Directory: C:/Dev/vcpkg/buildtrees/python3/x64-windows-rel
    Error code: 1
    See logs for more information:
      C:\Dev\vcpkg\buildtrees\python3\build-x64-windows-rel-out.log
> 
> Call Stack (most recent call first):
  scripts/cmake/vcpkg_build_msbuild.cmake:119 (vcpkg_execute_required_process)
  ports/python3/portfile.cmake:39 (vcpkg_build_msbuild)
  scripts/ports.cmake:94 (include)
> 
> Error: Building package python3:x64-windows failed with: BUILD_FAILED
Please ensure you're using the latest portfiles with `.\vcpkg update`, then
submit an issue at https://github.com/Microsoft/vcpkg/issues including:
  Package: python3:x64-windows
  Vcpkg version: 2020.01.17-nohash-external

Looking into build-x64-windows-rel-out.log shows a few warnings, for example
> C:\Dev\vcpkg\buildtrees\python3\src\v3.7.3-1161ad7a91-Lib-Win\PCBuild\pythoncore.vcxproj(457,5): warning : Toolset v142 is not used for official builds. Your build may have errors or incompatibilities.
     C:\Dev\vcpkg\buildtrees\python3\src\v3.7.3-1161ad7a91-Lib-Win\PCBuild\pythoncore.vcxproj(460,5): warning : Not including zlib is not a supported configuration.

but especially 72 times this error:
> C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\shared\sdkddkver.h(291,1): fatal error C1189: #error:  NTDDI_VERSION setting conflicts with _WIN32_WINNT setting

Because of the warning "Toolset v142 is not used for official builds. Your build may have errors or incompatibilities." I installed Toolset 14.1 and added
> set(VCPKG_PLATFORM_TOOLSET "v141")

to the portfile, but that only made that specific warning go away.

Does anyone else have this issue?

**Comment by NancyLi1013** [comment]
@jbmartens 
Could you please help confirm if the issue you encountered is the same as that before?
In my opinion, the old issue is caused by installing python2 and python3 in the same time.
And now this issue has been fixed. 
As for the issue you encountered, it seems that it is related with Visual Studio ToolSet version.
Please double check about this.
If it is still a problem for you, please open a new issue.

**Comment by JackGoldfinch** [comment]
@NancyLi1013 
I do believe that it is the same issue.

If I read the first post correctly, in order to install mlpack, first boost-python and of course python3 itself have to be built and installed. The whole process starts with package 1 of 14: python3[core]. And that's the build/installation, that fails.

It would probably be very helpful to see, if @jbmartens' VCPKG\buildtrees\python3\build-x64-windows-rel-out.log contains
> fatal error C1189: #error: NTDDI_VERSION setting conflicts with _WIN32_WINNT setting

as well.

Reading the whole thread again, I notice that manually adding the include path of python3 to boost-python didn't work, because - as @jbmartens states:
> I noticed that the directory  C:/vcpkg/installed/x64-windows/include/python3.7 doesn’t seem to exist.

I'm wondering, why would it exist? The build and installation of python3 failed...

I did a test just now, after updating vcpkg to Version 2020.02.04-nohash-external, the commit is [c5f01e1]:
I removed and purged all installed packages and deleted all remaining downloads and buildtrees as well. Then I tried to

`vcpkg install python3`

After successfully building and installing the dependency packages
> (1/4) libffi
(2/4) openssl-windows
(3/4) openssl

building package (4/4) python3 fails with:

> CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:72 (message):
    Command failed: msbuild C:/Dev/vcpkg/buildtrees/python3/src/v3.7.3-1161ad7a91-Lib-Win/PCBuild/pythoncore.vcxproj /p:Configuration=Release /t:Rebuild /p:Platform=x64 /p:PlatformToolset=v142 /p:VCPkgLocalAppDataDisabled=true /p:UseIntelMKL=No /p:WindowsTargetPlatformVersion=10.0.18362.0 /m
    Working Directory: C:/Dev/vcpkg/buildtrees/python3/x64-windows-rel
    Error code: 1
    See logs for more information:
      C:\Dev\vcpkg\buildtrees\python3\build-x64-windows-rel-out.log
>
> Call Stack (most recent call first):
  scripts/cmake/vcpkg_build_msbuild.cmake:119 (vcpkg_execute_required_process)
  ports/python3/portfile.cmake:46 (vcpkg_build_msbuild)
  scripts/ports.cmake:90 (include)
>
> Error: Building package python3:x64-windows failed with: BUILD_FAILED
Please ensure you're using the latest portfiles with '.\vcpkg update', then
submit an issue at https://github.com/Microsoft/vcpkg/issues including:
  Package: python3:x64-windows
  Vcpkg version: 2020.02.04-nohash-external
>
>Additionally, attach any relevant sections from the log files above.`

C:\Dev\vcpkg\buildtrees\python3\build-x64-windows-rel-out.log includes lots of
> C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\shared\sdkddkver.h(291,1): fatal error C1189: #error:  NTDDI_VERSION setting conflicts with _WIN32_WINNT setting [C:\Dev\vcpkg\buildtrees\python3\src\v3.7.3-1161ad7a91-Lib-Win\PCBuild\pythoncore.vcxproj]

and no other errors, ie the same as before.

So long story short: I'm pretty sure it's the same issue, as in my opinion the old issue was actually not about about a failing build of mlpack (or boost-python) due to python 2 and 3 being installed at the same time, but because the build and installation of python3 itself failed/fails.
If preferable, though, I'll gladly open a separate issue.

**Comment by NancyLi1013** [comment]
@ JackGoldfinch
Thanks for your so detailed info and deep investigation about this issue.
I will try it again on my local with clean environment.

Note: If this is the same issue, we continue to track this here.

**Comment by NancyLi1013** [comment]
Hi @JackGoldfinch 
Sorry for the late reply.
I used another vcpkg repo in a clean environment to build `python2 `and `python3` at the same time. There is no any error.

Could you please try to update vcpkg to the latest version to rebuild this? Since `python3 `has been updated to 3.8.3.

**Comment by JackGoldfinch** [comment]
Yes! It's working now! Thank you very much!
